### PR TITLE
add check validity of edm::Handle in HLTDQM modules [forwardport of #24445]

### PR DESCRIPTION
**Pull request by mtosi**
forwardport of https://github.com/cms-sw/cmssw/pull/24445

### CODE REVIEW

**Comment by mtosi** [comment]
@cmsbuild, please test

**Comment by fabiocos** [comment]
@jfernan2 @andrius-k @schneiml please check and sign it in case. The real solution looks to me adding the appropriate sequence to have the needed collection in the DQM sequence

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 1]
```diff
@@ -364,13 +374,40 @@ void BPHMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
...
   edm::Handle<reco::TrackCollection> trHandle;
   iEvent.getByToken( trToken_, trHandle );
+  if ( !trHandle.isValid() ) {
+    if (!warningPrinted4token_[2]) {
+      edm::LogWarning("BPHMonitor") << "skipping events because the collection " << trInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[2] = true;
+    }
+    return;
+  }
+  if ( !trHandle.isValid() ) {
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: @mtosi why a second time? Is it a double cut-and-paste? I am missing the logic here...

**Comment by fabiocos** [comment]
please test

File: DQMOffline/Trigger/plugins/HTMonitor.cc [hunk 2]
```diff
@@ -201,36 +229,57 @@ void HTMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)
...
-    if ( eleSelection_( e ) ) electrons.push_back(e);
+  if ( eleHandle.isValid() ) {
+    if ( eleHandle->size() < nelectrons_ ) return;
+    for ( auto const & e : *eleHandle ) {
+      if ( eleSelection_( e ) ) electrons.push_back(e);
+    }
+    if ( electrons.size() < nelectrons_ ) return;
+  } else {
+    if (!warningPrinted4token_[2]) {
+      edm::LogWarning("HTMonitor") << "skipping events because the collection " << jetInputTag_.label().c_str() << " is not available";
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/HTMonitor.cc]: hi-
I generally find these changes distasteful - however one practical comment - there is nothing here that actually skips the event. instead any checks on the number of good electrons are just skipped Probably that is not the behavior desired. 
(note also that the variable electrons is completely unnecessary it seems)

Comment (fabiocos) [reply DQMOffline/Trigger/plugins/HTMonitor.cc]: @davidlange6 as discussed directly we agree, I consider this as a quick workaround, the final solution is to properly add the producer in the correct sequence

File: DQMOffline/Trigger/plugins/BPHMonitor.cc (hunk 1 above)
Comment (mtosi) [reply DQMOffline/Trigger/plugins/BPHMonitor.cc]: :(

File: DQMOffline/Trigger/plugins/HTMonitor.cc (hunk 2 above)
Comment (mtosi) [reply DQMOffline/Trigger/plugins/HTMonitor.cc]: if the input electron collection is wrong, because not available, I doubt one can rely on any outcome of this code
=> cutting or not on the electrons multiplicity loses its value, I think
said that, I'm almost sure not all the consumers / users of this module need electrons,
=> I preferred to keep a conservative approach, which allows to process the event

**Comment by mtosi** [comment]
@cmsbuild, please test

**Comment by davidlange6** [comment]
please test

File: DQMOffline/Trigger/plugins/HTMonitor.cc (hunk 2 above)
Comment (fabiocos) [reply DQMOffline/Trigger/plugins/HTMonitor.cc]: @mtosi but you would need a return statement after setting the flag to true, right?

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 3]
```diff
@@ -658,25 +688,31 @@ void BPHMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
...
+		  phPhi_.numerator->Fill(p.phi(),PrescaleWeight);
+		  phEta_.numerator->Fill(p.eta(),PrescaleWeight);
+		  phPt_.numerator ->Fill(p.pt(),PrescaleWeight);
+		}		
 	      }
-              
+	  } else {
+	    if (!warningPrinted4token_[3]) {
+	      edm::LogWarning("BPHMonitor") << "skipping events because the collection " << phInputTag_.label().c_str() << " is not available";
+	      warningPrinted4token_[3] = true;
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: @mtosi don't you need a return statement here?

File: DQMOffline/Trigger/plugins/HTMonitor.cc [hunk 4]
```diff
@@ -201,36 +229,57 @@ void HTMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)
...
+      
+      if (!isFake) {
+	vtx = v;
+	break;
+      }
+    }
+  } else {
+    if (!warningPrinted4token_[3]) {
+      edm::LogWarning("HTMonitor") << "skipping events because the collection " << vtxInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[3] = true;
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/HTMonitor.cc]: @mtosi and a return statement here?

File: DQMOffline/Trigger/plugins/HTMonitor.cc [hunk 5]
```diff
@@ -201,36 +229,57 @@ void HTMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)
...
+    if ( muoHandle->size() < nmuons_ ) return;
+    for ( auto const & m : *muoHandle ) {
+      if ( muoSelection_( m ) && m.isGlobalMuon() && m.isPFMuon() && m.globalTrack()->normalizedChi2() < 10. && m.globalTrack()->hitPattern().numberOfValidMuonHits() > 0 && m.numberOfMatchedStations() > 1 && fabs(m.muonBestTrack()->dxy(vtx.position())) < 0.2 && fabs(m.muonBestTrack()->dz(vtx.position())) < 0.5 && m.innerTrack()->hitPattern().numberOfValidPixelHits() > 0 && m.innerTrack()->hitPattern().trackerLayersWithMeasurement() > 5 )  muons.push_back(m);
+    }
+    if ( muons.size() < nmuons_ ) return;
+  } else {
+    if (!warningPrinted4token_[4]) {
+      edm::LogWarning("HTMonitor") << "skipping events because the collection " << muoInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[4] = true;
+    }
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/HTMonitor.cc]: @mtosi and here...

File: DQMOffline/Trigger/plugins/METMonitor.cc [hunk 6]
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
...
+    if ( jetHandle->size() < njets_ ) return;
+    for ( auto const & j : *jetHandle ) {
+      if ( jetSelection_(j) ) {
+	jets.push_back(j);
+      }
+    }
+  } else {
+    if (!warningPrinted4token_[1]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << jetInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[1] = true;
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/METMonitor.cc]: @mtosi and here

File: DQMOffline/Trigger/plugins/METMonitor.cc [hunk 7]
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
...
+  if ( eleHandle.isValid() ) {
+    if ( eleHandle->size() < nelectrons_ ) return;
+    for ( auto const & e : *eleHandle ) {
+      if ( eleSelection_( e ) ) electrons.push_back(e);
+    }
+    if (electrons.size() < nelectrons_ ) return;
+  } else {
+    if (!warningPrinted4token_[2]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << jetInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[2] = true;
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/METMonitor.cc]: @mtosi and here

File: DQMOffline/Trigger/plugins/METMonitor.cc [hunk 8]
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
...
-      break;
+      if (!isFake) {
+	vtx = v;
+	break;
+      }
+    }
+  } else {
+    if (!warningPrinted4token_[3]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << vtxInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[3] = true;
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/METMonitor.cc]: @mtosi and here

File: DQMOffline/Trigger/plugins/METMonitor.cc [hunk 9]
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
...
+    if ( muoHandle->size() < nmuons_ ) return;
+    for ( auto const & m : *muoHandle ) {
+      bool pass = m.isGlobalMuon() && m.isPFMuon() && m.globalTrack()->normalizedChi2() < 10. && m.globalTrack()->hitPattern().numberOfValidMuonHits() > 0 && m.numberOfMatchedStations() > 1 && fabs(m.muonBestTrack()->dxy(vtx.position())) < 0.2 && fabs(m.muonBestTrack()->dz(vtx.position())) < 0.5 && m.innerTrack()->hitPattern().numberOfValidPixelHits() > 0 && m.innerTrack()->hitPattern().trackerLayersWithMeasurement() > 5;
+      if ( muoSelection_( m ) && pass ) muons.push_back(m);
+    }
+    if ( muons.size() < nmuons_ ) return;
+  } else {
+    if (!warningPrinted4token_[4]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << muoInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[4] = true;
```
Comment (fabiocos) [inline DQMOffline/Trigger/plugins/METMonitor.cc]: @mtosi and here

File: L1Trigger/Configuration/python/SimL1Emulator_cff.py [hunk 10]
```diff
@@ -62,7 +62,8 @@
 _phase2_siml1emulator += hgcalTriggerPrimitives
 
 from Configuration.Eras.Modifier_phase2_hgcal_cff import phase2_hgcal
-phase2_hgcal.toReplaceWith( SimL1Emulator , _phase2_siml1emulator )
+from Configuration.Eras.Modifier_phase2_hgcalV9_cff import phase2_hgcalV9
```
Comment (fabiocos) [inline L1Trigger/Configuration/python/SimL1Emulator_cff.py]: this change looks unrelated to the rest of the PR, it is just #24418 already merged picked up

File: DQMOffline/Trigger/plugins/HTMonitor.cc (hunk 2 above)
Comment (mtosi) [reply DQMOffline/Trigger/plugins/HTMonitor.cc]: the selections on electron will not be done, but the plots will be (wrongly) filled ...
but, if one does not want to apply any cuts on the electrons multiplicity and indeed even misconfigured the module, the plot will be (properly) filled and fine, and the warning message will make people know to update the config

am I wrong ?

File: DQMOffline/Trigger/plugins/BPHMonitor.cc (hunk 3 above)
Comment (fabiocos) [reply DQMOffline/Trigger/plugins/BPHMonitor.cc]: sure, the code could still run, but it will give different results wrt the fully functional version. So depending on which input is missing you could get different results from the same code without immediately realising it (in the DQM checks, people will not normally look at log files).
It is a matter of choices, having no filled histogram is in my opinion perhaps preferable than having filled histograms filled in an unexpected way. But I think this is ultimately something that should be agreed by the end users. We are always speaking of pathological situations.

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
@mtosi when you update the code, could you please rebase it onto a recent IB so as to avoid spurious insertion of unrelated commits?

**Comment by mtosi** [comment]
@cmsbuild, please test

**Comment by mtosi** [comment]
@fabiocos could this PR be integrated and a patch release w/ this be built ?
TSG asked to PdmV to produce samples for the 102X validation w/ data, and we need this PR in

please, let me know if there is anything to do
and thanks !

**Comment by fabiocos** [comment]
@jfernan2 @andrius-k @schneiml could you please check and sing it in case? We need to integrate the backport in 10_2_X to fix the present problem in the release

**Comment by andrius-k** [comment]
Hi Mia,

It seems that changes in this PR and the original [backport](https://github.com/cms-sw/cmssw/pull/24445) don't match. Maybe you forgot to push one more commit here?

**Comment by fabiocos** [comment]
@andrius-k I let @mtosi comment better, but as far as I can see the modified code is already different on input, and it looks to me that the differences in changes adapt to this original difference. Do you see places where this is not true?

**Comment by mtosi** [comment]
HTMonitor.cc and METMonitor.cc are identical
while BPHMonitor is different in the 2 releases
=> I modified manually the 2 files
I introduced `vtxInputTag_` in the master, because there is the corresponding Token only here (even if it is not used by the code)
then, warningPrinted4token_ has a larger size in the master than 102X (see above comment)

the rest reflects the original differences between the master and 102X

@andrius-k are you referring to a special piece of code ?

thanks !

**Comment by andrius-k** [comment]
+1
I just noticed that the number of commits and code changes are significantly different between the two PRs and was just double checking that this is intended.

**Comment by fabiocos** [comment]
+1
